# UTS Artificial Intelligence

Artificial Neural Network (ANN) dengan Algoritma BackPropagation

Nama  : Muhammad Fauzi Arkan
NIM   : 09011281621048
Kelas : SK5B Indralaya

Pada kasus ini akan memprediksi bagaimana pengaruh dari jam istirahat dan jam belajar dari seorang siswa terhadap nilai tes/ujian nya.
Dalam data dibawah ditunjukkan bahwa :
1. Apabila siswa memiliki jam untuk istirahat selama 2 jam dan jam untuk belajar selama 9 jam, maka siswa tersebut akan mendapatkan nilai 92.
2. Apabila siswa memiliki jam untuk istirahat selama 1 jam dan jam untuk belajar selama 5 jam, maka siswa tersebut akan mendapatkan nilai 86.
3. Apabila siswa memiliki jam untuk istirahat selama 3 jam dan jam untuk belajar selama 6 jam, maka siswa tersebut akan mendapatkan nilai 89.

Dengan menggunakan Artificial Neural Network (ANN) dan Algoritma BackPropagation diharapkan agar mendapatkan prediksi yang lebih tepat (meminimalkan nilai eror) dengan melakukan perubahan - perubahan pada nilai bobotnya. Karena tujuan dari ANN adalah untuk menghasilkan nilai prediksi (output) yang semirip mungkin dengan y (nilai prediksi awal). Setelah mendapatkan nilai eror, maka kita akan mulai memperbaiki nya dengan menggunakan algoritma BackPropagation.

Rumus utama untuk memperbaiki suatu bobot (W) berdasarkan eror (E) adalah :

W_new = W_old - Alpha (learning rate) x Turunan Parsial E / Turunan Parsial W

Rumus tersebut diatas juga berlaku untuk memperbaiki nilai bias.

Referensi :
1. Neural Network using BackPropogation in Python (https://www.youtube.com/watch?v=7qYtIveJ6hU&t=14s)
2. Machine Learning Made Easy (https://github.com/geeksnome/machine-learning-made-easy/blob/master/backpropogation.py)
3. CONTOH PERHITUNGAN ALGORITMA BACKPROPAGATION (https://structilmy.com/2019/07/contoh-perhitungan-algoritma-backpropagation/)

In [31]:
import numpy as np

In [32]:
# X = (Jam Istirahat, Jam Belajar), y = Nilai tes/ujian dari siswa
X = np.array(([2, 9], [1, 5], [3, 6]), dtype=float)
y = np.array(([92], [86], [89]), dtype=float)

# scale units
X = X/np.amax(X, axis=0) #jumlah maksimum dari array X
y = y/100 # nilai maksimum untuk hasil tes/ujian adalah 100

print(X)
print(y)

[[0.66666667 1.        ]
 [0.33333333 0.55555556]
 [1.         0.66666667]]
[[0.92]
 [0.86]
 [0.89]]


In [33]:
class NeuralNetwork(object):
    def __init__(self):
        #parameter
        self.inputSize = 2
        self.outputSize = 1
        self.hiddenSize = 3
        
         #bobot
        self.W1 = np.random.randn(self.inputSize, self.hiddenSize) # (3x2) bobot matrix dari input ke hidden layer
        self.W2 = np.random.randn(self.hiddenSize, self.outputSize) # (3x1) bobot matrix dari hidden layer ke output layer
        
    def feedForward(self, X):
        #forward propogation through the network
        self.z = np.dot(X, self.W1) #dot produk dari X (input) dan set bobot pertama (3x2)
        self.z2 = self.sigmoid(self.z) #fungsi aktivasi
        self.z3 = np.dot(self.z2, self.W2) #dot produk dari hidden layer (z2) dan set bobot kedua (3x1)
        output = self.sigmoid(self.z3)
        return output

    def sigmoid(self, s, deriv=False):
        if (deriv == True):
            return s * (1 - s)
        return 1/(1 + np.exp(-s))
    
    def backward(self, X, y, output):
        #backward propogate through the network
        self.output_error = y - output # error in output
        self.output_delta = self.output_error * self.sigmoid(output, deriv=True)
        
        self.z2_error = self.output_delta.dot(self.W2.T) #z2 error: berapa banyak bobot hidden layer yang mempengaruhi eror dari output
        self.z2_delta = self.z2_error * self.sigmoid(self.z2, deriv=True) #aplikasi turunan dari sigmoid untuk mengatasi eror pada z2
        
        self.W1 += X.T.dot(self.z2_delta) # menyesuaikan bobot set pertama (input -> hidden) 
        self.W2 += self.z2.T.dot(self.output_delta) # menyesuaikan bobot set kedua (hidden -> output)
        
        
    def train(self, X, y):
        output = self.feedForward(X)
        self.backward(X, y, output)
        
NN = NeuralNetwork()

for i in range(1000): #melatih Neural Network (NN) sebanyak 1000 kali
    if (i % 100 == 0):
        print("Loss: " + str(np.mean(np.square(y - NN.feedForward(X)))))
    NN.train(X, y)
    
print("Input: " + str(X))
print("Actual Output: " + str(y))
print("Loss: " + str(np.mean(np.square(y - NN.feedForward(X)))))
print("\n")
print("Predicted Output: "+ str(NN.feedForward(X)*100))

Loss: 0.52732813887255
Loss: 0.00014085125889791601
Loss: 0.00012385613490973224
Loss: 0.00011987887141548592
Loss: 0.0001161148381406696
Loss: 0.0001125200335039746
Loss: 0.00010908390970640134
Loss: 0.00010579670385456579
Loss: 0.00010264931454132627
Loss: 9.963325930157956e-05
Input: [[0.66666667 1.        ]
 [0.33333333 0.55555556]
 [1.         0.66666667]]
Actual Output: [[0.92]
 [0.86]
 [0.89]]
Loss: 9.674063425172694e-05


Predicted Output: [[90.57110465]
 [86.88008872]
 [89.29312275]]


Dapat dilihat pada Prediksi Output diatas menghasilkan nilai :
1. Siswa yang memiliki waktu istirahat selama 2 jam dan waktu belajar selama 9 jam, kemungkinan mendapatkan nilai 90.57 dari prediksi awal yaitu 92.
2. Siswa yang memiliki waktu istirahat selama 1 jam dan waktu belajar selama 5 jam, kemungkinan mendapatkan nilai 86.88 dari prediksi awal yaitu 86.
3. Siswa yang memiliki waktu istirahat selama 3 jam dan waktu belajar selama 6 jam, kemungkinan mendapatkan nilai 89.29 dari prediksi awal yaitu 89.